docs/timeouts.rst

```diff
@@ -0,0 +1,180 @@
+you may disable them or use lower-level APIs if you want more control over the
+behavior.
+
+Built-in Client Timeouts
+------------------------
+
+The high-level client APIs :func:`open_websocket` and :func:`open_websocket_url`
+contain built-in timeouts for connecting to a WebSocket and disconnecting from a
+WebSocket. These timeouts are built-in for two reasons:
+
+1. Omitting timeouts may be dangerous, and this library strives to make safe
+   code easy to write.
+2. These high-level APIs are context managers, and composing timeouts with
+   context managers is tricky.
+
+These built-in timeouts make it easy to write a WebSocket client that won't hang
+indefinitely if the remote endpoint or network are misbehaving. The following
+example shows a connect timeout of 10 seconds. This guarantees that the block
+will start executing (reaching the line that prints "Connected") within 10
+seconds. When the context manager exits after the ``print(Received response:
+…)``, the disconnect timeout guarantees that it will take no more than 5 seconds
+to reach the line that prints "Disconnected". If either timeout is exceeded,
+then the entire block raises ``trio.TooSlowError``.
+
+.. code-block:: python
+
+    async with open_websocket_url('ws://my.example/', connect_timeout=10,
+            disconnect_timeout=5) as ws:
+        print("Connected")
+        await ws.send_message('hello from client!')
+        response = await ws.get_message()
+        print('Received response: {}'.format(response))
+    print("Disconnected")
+
+.. note::
+
+    The built-in timeouts do not affect the contents of the block! In this
+    example, the client waits to receive a message. If the server never sends a
+    message, then the client will block indefinitely on ``ws.get_message()``.
+    Placing timeouts inside blocks is discussed below.
+
+What if you decided that you really wanted to manage the timeouts yourself? The
+following example implements the same timeout behavior explicitly, without
+relying on the library's built-in timeouts.
+
+.. code-block:: python
+
+    with trio.move_on_after(10) as cancel_scope:
+        async with open_websocket_url('ws://my.example',
+                connect_timeout=math.inf, disconnect_timeout=math.inf):
+            print("Connected")
+            cancel_scope.deadline = math.inf
+            await ws.send_message('hello from client!')
+            response = await ws.get_message()
+            print('Received response: {}'.format(response))
+            cancel_scope.deadline = trio.current_time() + 5
+        print("Disconnected")
+
+Notice that the library's internal timeouts are disabled by passing
+``math.inf``. This example is less ergonomic than using the built-in timeouts.
+If you really want to customize this behavior, you may want to use the low-level
+APIs instead, which are discussed below.
+
+Timeouts Inside Blocks
+----------------------
+
+The built-in timeouts do not apply to the contents of the block. One of the
```

Comment:
Another thought on reordering:  start with this example, because it's a common use, and the library won't provide any API for it.   Then mention that these timeouts don't apply to connect/disconnect, and introduce the example of controlling those timeouts manually.  Then introduce the API's connect/disconnect timeouts.

Reply:
This is a great suggestion! I will reorder this doc a bit.